omegaconf/basecontainer.py

```diff
@@ -528,7 +530,10 @@ def assign(value_key: Any, value_to_assign: Any) -> None:
 
     @staticmethod
     def _item_eq(
-        c1: Container, k1: Union[str, int], c2: Container, k2: Union[str, int]
+        c1: Container,
+        k1: Union[DictKeyType, int],
```

Comment:
Just a note that I had thought about suggesting that as well, then I also thought that keeping the union is semantically clearer (we want either a dict key or an int -- that some day might get written `ListindexType`). Python simplifies it anyway so there's no redundancy in practice.

Reply:
Yeah, I was just scratching my head looking at the `Union[DictKeyType, int]`.

Reply:
A related example: BaseContainer._get_full_key has signature
`def _get_full_key(self, key: Union[str, Enum, int, slice, None]) -> str:`
whose union type could be simplified to either `Union[DictKeyType, int, slice, None]` or `Union[DictKeyType, slice, None]`, depending on which style we choose (either explicitly specifying int for semantic clarity, or excluding it because it's redundant with the `int` in DictKeyType).

Reply:
I guess we can also go with something like:
`Union[DicyKeyType, ListKeyType]` where
```
ListKeyType = Union[int, slice]
```

Reply:
> ListKeyType = Union[int, slice]

Not sure we want the slice in there: maybe this specific function would work with slices, but the `get` / `pop` / `insert` methods of `ListConfig` wouldn't. And if a `ListKeyType` is defined, it should probably be used in these methods too.

Reply:
Although this is not a documented feature, `ListConfig.get(slice)` and `ListConfig.pop(slice)` both work as expected.
```
>>> cfg = OmegaConf.create([0,1,2,3,4])
>>> cfg
[0, 1, 2, 3, 4]
>>> cfg.get(slice(1,3))
[1, 2]
>>> cfg.pop(slice(1,3))
[1, 2]
>>> cfg
[0, 3, 4]
```

Reply:
I guess it's not clear how `list_config.insert(slice, ...)` would be defined. So `ListConfig.insert` has stricter typing requirements than `ListConfig.__getindex__`.

Reply:
I guess the confusion is that `slice` represents a set of indices, not a single index, and that methods such as `__getitem__` or `pop` are conventionally overloaded to accept either a single index or a set of indices.

Reply:
Yeah, I realized slice is not going to play well there after making the suggestion.
I think for now we can park this discussion and keep things as is.

> Although this is not a documented feature, ListConfig.get(slice) and ListConfig.pop(slice) both work as expected.
I think this is accidental :), I doubt it's a tested feature.